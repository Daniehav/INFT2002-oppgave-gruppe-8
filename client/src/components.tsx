import * as React from 'react';
import { Component } from 'react-simplified';
import {Question, Answer, questionService, answerService} from './service'
import { createHashHistory } from 'history';

export class QuestionDetails extends Component<{ match: { params: { id: number } } }> {
    question: Question = { id: 0, text: '', answered: false, tags: []};

    render() {
        return <div>
            <h4>{this.question.text}</h4><h6>answered: {this.question.answered}</h6>
            <div id="tags"><div>tag1</div><div>tag2</div></div>
        </div>
    }
    mounted() {
        questionService.get(this.props.match.params.id).then((question: Question) => (this.question = question))
        questionService.getTags(this.props.match.params.id).then((tags: string[]) => (this.question.tags = tags))
    }
}
/*
export class SearchBar extends Component<{match: {params: {text : string}}}>{
    searchTerm: string = '';

    render() {
        return <div>
            <input type='text' onChange={(event)=>this.searchTerm = event.currentTarget.value} ></input>
            <button onClick={() => history.push('/q/')}>Search</button>
        </div>
    }
}
*/
