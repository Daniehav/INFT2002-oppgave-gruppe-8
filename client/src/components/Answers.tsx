import React, {useState, useEffect, useContext} from 'react';
import { Outlet, useNavigate, useLocation } from 'react-router-dom';
import {questionService, answerService, tagService, commentService, profileService, favoriteService} from '../service'
import { Question, Answer, Tag, QuestionComment, Profile, AnswerComment} from '../types'
import { useParams, Link } from 'react-router-dom';
import { ProfileContext, AuthContext } from '../context/Context';
import Pfp from './Pfp';

import sortDown from '../assets/sort-down.svg'
import sortUp from '../assets/sort-up.svg'
import downvote from '../assets/downvote.svg'
import upvote from '../assets/upvote.svg'
import accepted from '../assets/accepted.svg'
import favorite from '../assets/favorite.svg'
import favorited from '../assets/favorited.svg'
import { formatDate } from './Questions';
import {UsernamePfp} from './Questions'
import { Comments, CreateComment } from './Comments';

export function Answers({question, setShowCreateComment}: {question: Question, setShowCreateComment: React.Dispatch<React.SetStateAction<boolean>>}){

    const [answers, setAnswers] = useState<Answer[]>([])
    const {profile} = useContext(ProfileContext)
    const {isAuthenticated} = useContext(AuthContext)
    const [descending, setDescending] = useState(true)
    const [sortBy, setSortBy] = useState<'score' | 'latest'>('score')
    const [userHasVoted, setUserHasVoted] = useState<number[]>([])
    const [favorites, setFavorites] = useState<number[]>([])

    
    

    useEffect(() => {
        
        sortAnswers(answers, descending)
    }, [sortBy, descending]);

    const sortAnswers = (answers: Answer[], descending: boolean) => {
        let sorted;
    
        switch (sortBy) {
          case 'score':
            sorted = [...answers].sort((a, b) => (b.upvotes - b.downvotes) - (a.upvotes - a.downvotes));
            break;
          case 'latest':
            sorted = [...answers].sort((a, b) => new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime());
            break;
          default:
            sorted = answers;
        }
    
        if (!descending) {
          sorted.reverse();
        }
    
        setAnswers(sorted);
    };


    useEffect(() => {
        const fetch = async() => {
            if(!question.question_id) return
            try {
                const answers = await answerService.getAll(question.question_id)
                sortAnswers(answers, true)
                const favorites = await favoriteService.getFavoriteIds()
                setFavorites(favorites)
                setDescending(true)
            } catch (error) {
                console.error(error);
            }
        }
        fetch()
    }, [question]);

    

    const acceptAnswer = async(answer: Answer) => {
        if(question.user_id != profile.user_id) return;
        setAnswers(prev => {
            return prev.map(a => {
                if(a.answer_id == answer.answer_id) {
                    return {...answer, accepted: !answer.accepted}
                } else {
                    return a
                }
            })
        })
        await questionService.accept(question.question_id, answer.answer_id, answer.user_id)
    }
    const favoriteAnswer = async(answer: Answer) => {
        if(answer.user_id == profile.user_id) return;
        setFavorites(prev => [...prev, answer.answer_id])
        const id = await favoriteService.setFavorite(answer.answer_id)
    }

    const vote = async (answer: Answer, vote: 'upvote' | 'downvote') => {
        if(answer.user_id == profile.user_id || userHasVoted.includes(answer.answer_id)) return
        setUserHasVoted(prev => [...prev, answer.answer_id])
        setAnswers(prev => {
            return prev.map(a => {
                if(a.answer_id == answer.answer_id) {
                    return {...answer, upvotes: vote == 'upvote'? answer.upvotes+1 : answer.upvotes,  downvotes: vote == 'downvote'? answer.downvotes + 1 : answer.downvotes}
                } else {
                    return a
                }
            })
        })
        await answerService.vote(answer.answer_id, vote)
    }

    const removeAnswer = (answer: Answer) => {
        setAnswers(prev => prev.filter(a => a.answer_id != answer.answer_id))
    }

    const answersElements = answers.map((answer, i) => {
        const isFavorite = favorites.includes(answer.answer_id)
        return <div key={i} className='wide-75'>
            <AnswerDetails answer={answer} question={question} accept={acceptAnswer} vote={vote} favoriteAnswer={favoriteAnswer} isFavorite={isFavorite} removeAnswer={removeAnswer}/>
        </div>
    })

    
    return(
        <>
            <div className='wide-75'>
                <div className='card bg-white row'>
                    <p>{answers.length} answers</p>
                    <button className={`text-black ${sortBy == 'score'? 'select-underline' : ''}`} onClick={() => setSortBy('score')}>Sort by score</button>
                    <img className='icon-s pointer' onClick={() => setDescending(prev => !prev)} src={descending? sortDown : sortUp} alt={descending? 'desc' : 'asc'} />
                    <button className={`text-black ${sortBy == 'latest'? 'select-underline' : ''}`} onClick={() => setSortBy('latest')}>Sort by latest</button>
                    {profile.user_id != question.user_id && isAuthenticated &&<>
                        <Link to={`/question/${question.question_id}/answer/create`} className="button bg-light-grey text-black fs-3">Post answer</Link>
                        <button onClick={() => setShowCreateComment(true)} className="button bg-light-grey text-black fs-3">Post comment</button>
                    </>}
                </div>
            </div>
            {answersElements}
        </> 
            
    )
} 

type AnswerDetailsProps = {
    answer: Answer,
    question: Question, 
    vote: (answer: Answer, vote: 'upvote' | 'downvote') => void,
    accept: (answer: Answer) => void,
    favoriteAnswer: (answer: Answer) => void,
    isFavorite: boolean,
    removeAnswer: (answer: Answer) => void
}

export function AnswerDetails({answer, question, vote, accept, favoriteAnswer, isFavorite, removeAnswer}: AnswerDetailsProps) {

    const {profile} = useContext(ProfileContext)
    const {isAuthenticated} = useContext(AuthContext)
    const [showCreateAnswerComment, setShowCreateComment] = useState<number | false>(false)
    const [comments, setComments] = useState<AnswerComment[]>([])

    const navigate = useNavigate()

    useEffect(() => {
        if(!answer.answer_id) return
        const fetch = async () => {
            const comments = await commentService.get('answer',answer.answer_id)
            setComments(comments as AnswerComment[])
        }
        fetch()
    }, []);

    const addComment = (comment: string, commentId: number) => {
        if(!comment || !commentId) return
        setShowCreateComment(false)
        const answerComment = {body: comment, comment_id: commentId, user_id: profile.user_id, answer_id: answer.answer_id} as AnswerComment
        setComments(prev => [...prev, answerComment])
    } 
    const editComment = (comment: string, commentId: number) => {
        if(!comment || !commentId) return
        setComments(prev => prev.map(c => {
            if(c.comment_id == commentId){
                return {...c, body: comment}
            } else {
                return c
            }
        }))
    } 
    //removes from state to update component
    const deleteComment = (commentId: number) => {
        if(!commentId) return
        setComments(prev => prev.filter((c) => c.comment_id != commentId)) 
    } 


    const postedQuestion = question.user_id == profile.user_id
    const postedAnswer = answer.user_id == profile.user_id 
    let acceptIconClass = answer.accepted? 'accepted' : postedQuestion? 'accept' : 'vis-hide'
    const favoriteIcon = isFavorite? favorited : favorite

    const deleteAnswer = async (answerId: number) => {
        removeAnswer(answer)
        await answerService.delete(answerId)
        navigate('/question/'+ question.question_id)
    }

    return ( <>
        <div className='card bg-white wide-100'>
            <div className='row'>
                <UsernamePfp userId={answer.user_id} withPfp={true} />
                {!postedAnswer && <img onClick={() => favoriteAnswer(answer)} className={`icon-m pointer`} src={favoriteIcon} alt={isFavorite? 'favorited' : 'not-favorited'}/>}
                <img onClick={() => accept(answer)} className={`icon-m ${acceptIconClass} ${postedQuestion? 'pointer' : ''}`} src={accepted} alt={answer.accepted? 'accepted' : 'not-accepted'}/>
                <p className='fs-5'>Asked {formatDate(answer.created_at)}</p>
                <p className='fs-5'>Modified {formatDate(answer.updated_at)}</p>
            </div>
            <div className="row space-between">
                <div className='row'>
                    <div className='flex-vert align-center gap-05'>
                        {!postedAnswer && <img onClick={() => vote(answer, 'upvote')} className='icon-s pointer' src={upvote} alt="upvote" />}
                        <p className='fs-3 vote-count'>{answer.upvotes - answer.downvotes}</p>
                        {!postedAnswer && <img onClick={() => vote(answer, 'downvote')} className='icon-s pointer' src={downvote} alt="downvote" />}
                    </div>
                    <p className='text-body'>{answer.body}</p>
                </div>
                <div className="flex-vert align-start gap-1">
                    {answer.user_id != profile.user_id && isAuthenticated && <button onClick={() => setShowCreateComment(answer.answer_id)} className="button bg-light-grey text-black fs-3">Post comment</button>}
                </div>
            </div>
            <div>
                <Comments comments={comments} editComment={editComment} removeComment={deleteComment} parent={'answer'} />
            </div>
            {postedAnswer && 
            <div className='row justify-end'>
                <Link className='button bg-light-grey text-black fs-4' to={`/question/${question.question_id}/answer/${answer.answer_id}/edit`}>Edit</Link>
                <button className='button bg-light-grey text-black fs-4' onClick={() => deleteAnswer(answer.answer_id)}>Delete</button>
            </div>}
        </div>
        {showCreateAnswerComment == answer.answer_id && <CreateComment parent='answer' parentId={answer.answer_id} addComment={addComment} show={setShowCreateComment} />}
    </>
)

}

export function CreateAnswer() {

    const [answer, setAnswer] = useState('')
    const navigate = useNavigate()
    const params = useParams()

    const post = async () => {
        if(!answer || !params.id ) return
        const questionId = parseInt(params.id)
        const answerId = await answerService.create(questionId, answer)
        navigate('/question/'+ questionId)
    }

    return(
        <div className='card wide-75 bg-white row'>
            <textarea className='textarea textarea--answer wide-75' value={answer} onChange={e => setAnswer(e.currentTarget.value)}></textarea>
            <button className='button bg-accent text-WHITE fs-3' onClick={post}>Post answer</button>
        </div>
    )
}

export function EditAnswer() {
    const params = useParams()
    const answerId = parseInt(params.answerId as string)
    const id = parseInt(params.id as string)
    const [answer, setAnswer] = useState<Answer>({} as Answer)
    const navigate = useNavigate()

    useEffect(() => {
        if(!answerId) return
        const fetch = async () => {
            const answer = await answerService.get(answerId)
            setAnswer(answer)
        }
        fetch()
    }, [answerId]);

    const edit =async () => {
        if(!answer.body) return
        await answerService.edit(answer);
        navigate('/question/'+ id)
    }

    const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
        
        setAnswer(prev => {
            return { ...prev, body: e.target.value} as Answer
        })
    }

    return(
        <div className='card wide-75 bg-white row'>
            <textarea className='textarea textarea--answer wide-75' value={answer.body} onChange={handleChange}></textarea>
            <div className="flex-vert align-start gap-1">
                <button className='button bg-light-grey text-black fs-3' onClick={() => navigate('/question/'+id)}>Cancel</button>
                <button className='button bg-accent text-WHITE fs-3' onClick={edit}>Update answer</button>
            </div>
        </div>
    )
}

