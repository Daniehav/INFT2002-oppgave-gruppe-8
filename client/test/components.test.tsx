import * as React from 'react'
import { MemoryRouter, Router } from 'react-router-dom';
import { MemoryHistory, createMemoryHistory } from 'history';
import {render, screen} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import {AuthContext, ProfileContext, ThemeContext} from '../src/context/Context'
import { Profile, Question, Tag, Answer, QuestionComment, AnswerComment } from '../src/types';
import { answerService, commentService, favoriteService, questionService, tagService } from '../src/service';
import { CreateQuestion, QuestionDetails } from '../src/components/Questions';
import { AnswerDetails, Answers } from '../src/components/Answers';
import { Comments } from '../src/components/Comments';
import { Searchbar } from '../src/components/Searchbar';
import { Tags } from '../src/Home';

const testQuestions  = [
    {question_id:  1, user_id: 3, title: 'hvilken bokstav kommer etter a i alfabetet', body: 'jeg prøver å lære meg en viss sang, men har glemt teksten', views: 2, created_at: new Date(), updated_at: new Date(),tags:[2]},
    {question_id:  2, user_id: 2, title: 'hva er 2+2?', body: 'jeg har allerede prøvd å gjette meg fram til svaret, har noen en bedre løsning?', views: 1, created_at: new Date(), updated_at: new Date(),tags:[1]},
    {question_id:  3, user_id: 2, title: 'er fluesopp spiselig?', body: 'jeg gikk ut ifra at de var det ettersom de ser så fargerike ut, men jeg begynner å føle meg litt rar', views: 3, created_at: new Date(), updated_at: new Date(),tags:[3]},
    {question_id:  4, user_id: 1, title: 'regne ut antall epler', body: 'hvis jeg har tre epler, og jeg får to til, hvor mange epler har jeg da?', views: 0, created_at: new Date(), updated_at: new Date(),tags:[3]},
]

const testTags: Tag[] = [
    {tag_id: 1, name: 'matte', count: 1, created_at: new Date, updated_at: new Date},
    {tag_id: 2, name: 'alfabet', count: 0, created_at: new Date, updated_at: new Date},
    {tag_id: 3, name: 'mat', count: 0, created_at: new Date, updated_at: new Date},
];

const testAnswers: Answer[] = [
    {answer_id: 1, question_id: 1, user_id: 1, body: 'l er bokstaven som kommer etter a', upvotes: 5, downvotes: 3, accepted: false, created_at: new Date(), updated_at: new Date()},
    {answer_id: 2, question_id: 1, user_id: 2, body: 'tror b er neste bokstav', upvotes: 0, downvotes: 0, accepted: true, created_at: new Date(), updated_at: new Date()},
    {answer_id: 3, question_id: 3, user_id: 1, body: 'fluesopp er ikke spiselig, oppsøk legevakta øyeblikkelig', upvotes: 0, downvotes: 0, accepted: true, created_at: new Date(), updated_at: new Date()},
    {answer_id: 4, question_id: 2, user_id: 3, body: '22', upvotes: 0, downvotes: 0, accepted: false, created_at: new Date(), updated_at: new Date()},
];

const testQuestionComments: QuestionComment[] = [
    {comment_id: 1, question_id: 1, user_id: 1, body: 'a finnes ikke i alfabetet',  created_at: new Date(), updated_at: new Date()},
    {comment_id: 2, question_id: 1, user_id: 2, body: 'hjelp jeg har samme problem', created_at: new Date(), updated_at: new Date()},
    {comment_id: 3, question_id: 3, user_id: 1, body: 'vanligvis så står det på stammen om den er spiselig eller ikke', created_at: new Date(), updated_at: new Date()},
    {comment_id: 4, question_id: 2, user_id: 3, body: 'går ikke å regne fram, du må fortsette å gjette', created_at: new Date(), updated_at: new Date()},
];
const testAnswerComments: AnswerComment[] = [
    {comment_id: 1, answer_id: 1, user_id: 1, body: 'det er feil', created_at: new Date(), updated_at: new Date()},
    {comment_id: 2, answer_id: 1, user_id: 2, body: 'er du dum?', created_at: new Date(), updated_at: new Date()},
    {comment_id: 3, answer_id: 3, user_id: 1, body: 'oi det visste ikke jeg, takk', created_at: new Date(), updated_at: new Date()},
    {comment_id: 4, answer_id: 2, user_id: 3, body: 'tusen takk', created_at: new Date(), updated_at: new Date()},
];


jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(),
}))

const defaultProfile: Profile = {
    display_name: "",
    id: 0,
    user_id: 0,
    profile_picture: null, // Replace with an actual default picture URL
    bio: "",
    level: 0,
    points: 0,
    username: "",
  };

const customRender = (children: React.ReactNode, profile: Profile, history: MemoryHistory) => {
    const setProfile = jest.fn()
    const isAuthenticated = true
    const setIsAuthenticated = jest.fn()
    const logOut = jest.fn()
    return render(
        <Router location={history.location} navigator={history}>
            <ProfileContext.Provider value={{profile, setProfile}}>
                <AuthContext.Provider value={{isAuthenticated, setIsAuthenticated, logOut}}>
                    {children}
                </AuthContext.Provider>
            </ProfileContext.Provider>
        </Router>
    )
}

// Mock implementations for QuestionService
jest.mock('../src/service', () => ({
    questionService: {
      get: jest.fn(),
      delete: jest.fn(),
      create: jest.fn(),
      search: jest.fn(),
      getQuestionTags: jest.fn(),
      getQuestionsUser: jest.fn(),
      getPreview: jest.fn(),
      getFilter: jest.fn(),
      getTagFilter: jest.fn(),
      edit: jest.fn(),
      accept: jest.fn(),
    },
    favoriteService: {
        getFavorites: jest.fn(),
        getFavoriteIds: jest.fn(),
        setFavorite: jest.fn(),
    },
    answerService: {
        get: jest.fn(),
        getAll: jest.fn(),
        create: jest.fn(),
        edit: jest.fn(),
        delete: jest.fn(),
        vote: jest.fn(),
    },
    commentService: {
        get: jest.fn(),
        create: jest.fn(),
        edit: jest.fn(),
        delete: jest.fn(),
    },
    tagService: {
        getAll: jest.fn(),
        create: jest.fn(),
        getQuestionTags: jest.fn(),
        editQuestionTags: jest.fn(),
    },
    profileService: {
        get: jest.fn().mockImplementation(() => Promise.resolve(defaultProfile)),
        update: jest.fn(),
        getProfileByUsername: jest.fn(),
    }
    
  }));
  


const useParamsMock = jest.spyOn(require('react-router-dom'), 'useParams');

const getQuestionMock = jest.spyOn(questionService, 'get')
const deleteQuestionMock = jest.spyOn(questionService, 'delete')
const createQuestionMock = jest.spyOn(questionService, 'create')
const searchMock = jest.spyOn(questionService, 'search')
const getQuestionTagsMock = jest.spyOn(tagService, 'getQuestionTags')
const getAnswersMock = jest.spyOn(answerService, 'getAll')
const getCommentsMock = jest.spyOn(commentService, 'get')
const getFavoritesMock = jest.spyOn(favoriteService, 'getFavoriteIds')
const voteAnswerMock = jest.spyOn(answerService,'vote')
const acceptAnswerMock = jest.spyOn(questionService,'accept')
const getAllTagsMock = jest.spyOn(tagService, 'getAll')

getCommentsMock.mockImplementation(() => Promise.resolve([]))
getFavoritesMock.mockImplementation(() => Promise.resolve([]))
deleteQuestionMock.mockImplementation(() => Promise.resolve())
voteAnswerMock.mockImplementation(() => Promise.resolve())
acceptAnswerMock.mockImplementation(() => Promise.resolve())




describe('Question components tests', () => {
    test('QuestionDetails renders question', async() => {
        useParamsMock.mockReturnValue({ id: 1 });
        const history = createMemoryHistory()
        getQuestionMock.mockImplementation(() => Promise.resolve(testQuestions[0]))
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))


        const questionTags = testQuestions[0].tags.map(i => testTags[i-1])
        getQuestionTagsMock.mockImplementation(() => Promise.resolve(questionTags))
        const profile = {user_id: 1}
        
        const {asFragment, findByText } = customRender(<QuestionDetails />, profile as Profile, history)
        await findByText(testQuestions[0].title)
        await findByText(testQuestions[0].body)
        for (const tag of questionTags) {
            await findByText(tag.name)
        }
        expect(asFragment()).toMatchSnapshot()
        
    })
    
    test('Render only what question creator has permission to', async() => {
        useParamsMock.mockReturnValue({ id: 1 });
        const history = createMemoryHistory()
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))


        const { findByText } = customRender(<QuestionDetails />, {user_id: 3} as Profile, history)
        await findByText('Edit')
        await findByText('Delete')
        expect(screen.queryByText('Post Answer')).not.toBeInTheDocument()
        expect(screen.queryByText('Post Comment')).not.toBeInTheDocument()
    })
    
    test('Delete question', async() => {
        useParamsMock.mockReturnValue({id:4})
        const history = createMemoryHistory()
        history.push('/question/4')

        const answers = testAnswers.filter(a => a.question_id == 4)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))


        getQuestionMock.mockImplementation(() => Promise.resolve(testQuestions[3]))
        const { findByText } = customRender(<QuestionDetails />, {user_id: 1} as Profile, history)
        const user = userEvent.setup()
        expect(history.location.pathname).toBe('/question/4')
        const deleteButton = await findByText('Delete')
        await user.click(deleteButton)
        expect(history.location.pathname).toBe('/')
    })
    
    test('Searching for questions and navigate to question', async () => {
        useParamsMock.mockReturnValue({ id: 1 });
        const history = createMemoryHistory()
        getQuestionMock.mockImplementation(() => Promise.resolve(testQuestions[0]))
        searchMock.mockImplementation(() => Promise.resolve([testQuestions[0], testQuestions[1]]))
        const {findByText} = customRender(<Searchbar />, {user_id: 1} as Profile, history)
        const user = userEvent.setup()
        const searchInput = screen.getByLabelText('search')
        await user.type(searchInput,'hv')
        const button = await findByText(testQuestions[0].title)
        await findByText(testQuestions[1].title)
        await user.click(button)
        expect(history.location.pathname).toBe('/question/1')
    })
    
    test('Searching for questions and navigate to results', async () => {
        useParamsMock.mockReturnValue({ id: 1 });
        const history = createMemoryHistory()
        getQuestionMock.mockImplementation(() => Promise.resolve(testQuestions[0]))
        const {findByText} = customRender(<Searchbar />, {user_id: 1} as Profile, history)
        const user = userEvent.setup()
        const searchInput = screen.getByLabelText('search')
        await user.type(searchInput,'hv')
        await findByText(testQuestions[0].title)
        await findByText(testQuestions[1].title)
        await user.keyboard('[Enter]')
        expect(history.location.pathname).toBe('/question/search/hv/results')
    })

    test('Create question', async() => {
        const history = createMemoryHistory()
        history.push('/question/create')
        getAllTagsMock.mockImplementation(() => Promise.resolve(testTags))
        createQuestionMock.mockImplementation(() => Promise.resolve('5'))
        const user = userEvent.setup()
        const {getByLabelText, findByText, getByText} = customRender(<CreateQuestion />, {user_id: 2} as Profile, history)
        const title = getByLabelText('title')
        await user.type(title,'Hva er 1+1')
        const body = getByLabelText('body')
        await user.type(body,'ikke noe mer å legge til')
        const tagInput = getByLabelText('tag-input')
        await user.type(tagInput,'mat')
        const matteOption = await findByText('matte')
        await user.click(matteOption)
        expect(title).toHaveValue('Hva er 1+1')
        expect(body).toHaveValue('ikke noe mer å legge til')
        await findByText('matte', {classname: 'tag-option'})
        await user.click(getByText('Post'))
        expect(history.location.pathname).toBe('/question/5')

    })
})

describe('Answer components tests', () => {
    test('AnswerDetails renders correctly', async () => {
        useParamsMock.mockReturnValue({ id: 1 });
        const history = createMemoryHistory()

        const { findByText } = customRender(<AnswerDetails question={testQuestions[0]} answer={testAnswers[0]} vote={jest.fn()} accept={jest.fn()} favoriteAnswer={jest.fn()} isFavorite={false} removeAnswer={jest.fn()} />, {user_id: 1} as Profile, history)
        await findByText(testAnswers[0].body)
        await findByText(testAnswers[0].upvotes - testAnswers[0].downvotes, {selector: '.vote-count'})
    })
    
    test('Answers count is correct', async () => {
        useParamsMock.mockReturnValue({id: 1})
        const history = createMemoryHistory()
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))

        const { findByText } = customRender(<Answers question={testQuestions[0]} setShowCreateComment={jest.fn()} />, {user_id: 1} as Profile, history)
        await findByText('2 answers')
        await findByText(answers[0].body)
    })
    
    test('Upvoting answer increments score', async() => {
        useParamsMock.mockReturnValue({id: 1})
        const history = createMemoryHistory()
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))

        const user = userEvent.setup()

        const { findByText } = customRender(<Answers question={testQuestions[0]} setShowCreateComment={jest.fn()} />, {user_id: 2} as Profile, history)
        const score = testAnswers[0].upvotes - testAnswers[0].downvotes
        await findByText(answers[0].body)
        const upvoteButton = screen.getAllByAltText('upvote')[0]
        await user.click(upvoteButton)
        await findByText(score + 1, {selector: '.vote-count'})
    })

    test('Downvoting answer decrements score', async() => {
        useParamsMock.mockReturnValue({id: 1})
        const history = createMemoryHistory()
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))

        const user = userEvent.setup()

        const { findByText } = customRender(<Answers question={testQuestions[0]} setShowCreateComment={jest.fn()} />, {user_id: 2} as Profile, history)
        const score = testAnswers[0].upvotes - testAnswers[0].downvotes
        await findByText(answers[0].body)
        const upvoteButton = screen.getAllByAltText('downvote')[0]
        await user.click(upvoteButton)
        await findByText(score - 1, {selector: '.vote-count'})
    })
    test('Accepting answer marks it as accepted', async() => {
        useParamsMock.mockReturnValue({id: 1})
        const history = createMemoryHistory()
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))

        const user = userEvent.setup()

        const { findByText } = customRender(<Answers question={testQuestions[0]} setShowCreateComment={jest.fn()} />, {user_id: 3} as Profile, history)
        await findByText(answers[0].body)
        const acceptButton = screen.getAllByAltText('not-accepted')[0]
        await user.click(acceptButton)
        expect(acceptButton).toHaveAttribute('alt','accepted')
    })
    test('Favoriting answer marks it as favorited', async() => {
        useParamsMock.mockReturnValue({id: 1})
        const history = createMemoryHistory()
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))
        getFavoritesMock.mockImplementation(() => Promise.resolve([]))
        acceptAnswerMock.mockImplementation(() => Promise.resolve())
        getCommentsMock.mockImplementation(() => Promise.resolve([]))

        const user = userEvent.setup()

        const { findByText } = customRender(<Answers question={testQuestions[0]} setShowCreateComment={jest.fn()} />, {user_id: 2} as Profile, history)
        await findByText(answers[0].body)
        const favoriteButton = screen.getAllByAltText('not-favorited')[0]
        await user.click(favoriteButton)
        expect(favoriteButton).toHaveAttribute('alt','favorited')
    })
    
    test('Sort by score (desc and asc)', async () => {
        useParamsMock.mockReturnValue({id: 1})
        const history = createMemoryHistory()
        const answers = testAnswers.filter(a => a.question_id == 1)
        getAnswersMock.mockImplementation(() => Promise.resolve(answers))
        const user = userEvent.setup()
    
        const { findByText, findByAltText, findAllByText } = customRender(<Answers question={testQuestions[0]} setShowCreateComment={jest.fn()} />, {user_id: 2} as Profile, history)
        const sort = await findByText('Sort by score')
        user.click(sort)
        const answerScoresDesc = await findAllByText(/^\d+$/,{selector:'.vote-count', exact: false})
        const firstDesc = parseInt(answerScoresDesc[0].textContent)        
        const secondDesc = parseInt(answerScoresDesc[1].textContent)        
        expect(secondDesc).toBeLessThan(firstDesc)
        const desc = await findByAltText('desc')
        await user.click(desc)
        const answerScoresAsc = await findAllByText(/^\d+$/,{selector:'.vote-count', exact: false})
        const firstAsc = parseInt(answerScoresAsc[0].textContent)        
        const secondAsc = parseInt(answerScoresAsc[1].textContent)        
        expect(desc).toHaveAttribute('alt','asc')
        expect(firstAsc).toBeLessThan(secondAsc)

        

    })
    
})

describe('Comment components tests', () => {
    test('Renders question comments correctly', async () => {
        useParamsMock.mockReturnValue({ id: 1 });
        const history = createMemoryHistory()
        const comments = testQuestionComments.filter(c => c.question_id == 1)
        getCommentsMock.mockImplementation(() => Promise.resolve(comments))
        const {findByText} = await customRender(<Comments comments={comments} parent='question' removeComment={jest.fn()} editComment={jest.fn()} />, {user_id: 1} as Profile, history)
        await findByText(testQuestionComments[0].body)
        await findByText(testQuestionComments[1].body)
    })
    test('Renders answer comments correctly', async () => {
        useParamsMock.mockReturnValue({ id: 1 });
        const history = createMemoryHistory()
        const comments = testAnswerComments.filter(c => c.answer_id == 1)
        getCommentsMock.mockImplementation(() => Promise.resolve(comments))
        const {findByText} = await customRender(<Comments comments={comments} parent='answer' removeComment={jest.fn()} editComment={jest.fn()} />, {user_id: 1} as Profile, history)
        await findByText(testAnswerComments[0].body)
        await findByText(testAnswerComments[1].body)
    })
})

describe('Other tests', () => {
    test('Filtering tags on name', async() => {
        getAllTagsMock.mockImplementation(() => Promise.resolve(testTags))
        const user = userEvent.setup()
        const {findByText} = customRender(<Tags />, {user_id: 1} as Profile, createMemoryHistory())
        await findByText('alfabet')
        const input = screen.getByLabelText('tag-input')
        await user.type(input, 'ma')
        await findByText(testTags[0].name)
        await findByText(testTags[2].name)
        expect(screen.queryByText('alfabet')).not.toBeInTheDocument()
    })
})
