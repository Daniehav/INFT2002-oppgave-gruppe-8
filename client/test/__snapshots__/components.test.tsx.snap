// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question components tests QuestionDetails renders question 1`] = `
<DocumentFragment>
  <div
    class="question-page"
  >
    <div
      class="wide-75 card bg-white"
    >
      <div
        class="question-header"
      >
        <div
          class="row"
        >
          <div
            class="row gap-1 pointer"
          >
            <div
              class="pfp-level"
            >
              <img
                alt=""
                class="pfp pfp-s"
                src="[object Object]"
              />
              <p
                class="level fs-5 text-WHITE"
              />
            </div>
            <p
              class="fs-4"
            />
          </div>
        </div>
        <p
          class="fs-1"
        >
          hvilken bokstav kommer etter a i alfabetet
        </p>
        <div
          class="row"
        >
          <p
            class="fs-5"
          >
            Asked Just now
          </p>
          <p
            class="fs-5"
          >
            Modified Just now
          </p>
          <p
            class="fs-5"
          >
            Viewed 2 times
          </p>
          <div
            class="tags"
          >
            <div
              class="tag fs-5"
            >
              alfabet
            </div>
          </div>
        </div>
      </div>
      <p
        class="fs-4 text-body"
      >
        jeg prøver å lære meg en viss sang, men har glemt teksten
      </p>
    </div>
    <div
      class="card bg-white wide-75"
    >
      <p
        class="fs-4"
      >
        0 Comments
      </p>
    </div>
    <div
      class="wide-75"
    />
    <div
      class="wide-75"
    >
      <div
        class="card bg-white row"
      >
        <p>
          0 answers
        </p>
        <button
          class="text-black select-underline"
        >
          Sort by score
        </button>
        <img
          alt="desc"
          class="icon-s pointer"
          src="[object Object]"
        />
        <button
          class="text-black "
        >
          Sort by latest
        </button>
        <a
          class="button bg-light-grey text-black fs-3"
          href="/question/1/answer/create"
        >
          Post answer
        </a>
        <button
          class="button bg-light-grey text-black fs-3"
        >
          Post comment
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
